File: stdlib/cruntime/math.c

#include <math.h>

// Constants
double dex_math_pi() {
    return 3.14159265358979323846;
}

double dex_math_e() {
    return 2.71828182845904523536;
}

// Trigonometric
double dex_math_sin(double x) {
    return sin(x);
}

double dex_math_cos(double x) {
    return cos(x);
}

double dex_math_tan(double x) {
    return tan(x);
}

double dex_math_asin(double x) {
    return asin(x);
}

double dex_math_acos(double x) {
    return acos(x);
}

double dex_math_atan(double x) {
    return atan(x);
}

// Power / root
double dex_math_sqrt(double x) {
    return sqrt(x);
}

double dex_math_pow(double base, double exp) {
    return pow(base, exp);
}

double dex_math_exp(double x) {
    return exp(x);
}

// Rounding
double dex_math_floor(double x) {
    return floor(x);
}

double dex_math_ceil(double x) {
    return ceil(x);
}

double dex_math_round(double x) {
    return round(x);
}

// Other
double dex_math_abs(double x) {
    return fabs(x);
}

double dex_math_log(double x) {
    return log(x);
}

double dex_math_log2(double x) {
    return log2(x);
}

double dex_math_log10(double x) {
    return log10(x);
}

// Two-arg
double dex_math_min(double a, double b) {
    return (a < b) ? a : b;
}

double dex_math_max(double a, double b) {
    return (a > b) ? a : b;
}

File: stdlib/math.go
package stdlib

import (
	_ "embed"

	"github.com/ensamuel7/dex/ast"
)

//go:embed cruntime/math.c
var mathRuntime string

func init() {
	Register(&Module{
		Name: "math",
		Funcs: map[string]FuncDef{
			// Constants (zero-arg, return double)
			"pi": {
				Params:     []ast.Type{},
				ReturnType: ast.TypeDouble,
				CName:      "dex_math_pi",
			},
			"e": {
				Params:     []ast.Type{},
				ReturnType: ast.TypeDouble,
				CName:      "dex_math_e",
			},
			// Trigonometric
			"sin": {
				Params:     []ast.Type{ast.TypeDouble},
				ReturnType: ast.TypeDouble,
				CName:      "dex_math_sin",
			},
			"cos": {
				Params:     []ast.Type{ast.TypeDouble},
				ReturnType: ast.TypeDouble,
				CName:      "dex_math_cos",
			},
			"tan": {
				Params:     []ast.Type{ast.TypeDouble},
				ReturnType: ast.TypeDouble,
				CName:      "dex_math_tan",
			},
			"asin": {
				Params:     []ast.Type{ast.TypeDouble},
				ReturnType: ast.TypeDouble,
				CName:      "dex_math_asin",
			},
			"acos": {
				Params:     []ast.Type{ast.TypeDouble},
				ReturnType: ast.TypeDouble,
				CName:      "dex_math_acos",
			},
			"atan": {
				Params:     []ast.Type{ast.TypeDouble},
				ReturnType: ast.TypeDouble,
				CName:      "dex_math_atan",
			},
			// Power / root
			"sqrt": {
				Params:     []ast.Type{ast.TypeDouble},
				ReturnType: ast.TypeDouble,
				CName:      "dex_math_sqrt",
			},
			"pow": {
				Params:     []ast.Type{ast.TypeDouble, ast.TypeDouble},
				ReturnType: ast.TypeDouble,
				CName:      "dex_math_pow",
			},
			"exp": {
				Params:     []ast.Type{ast.TypeDouble},
				ReturnType: ast.TypeDouble,
				CName:      "dex_math_exp",
			},
			// Rounding
			"floor": {
				Params:     []ast.Type{ast.TypeDouble},
				ReturnType: ast.TypeDouble,
				CName:      "dex_math_floor",
			},
			"ceil": {
				Params:     []ast.Type{ast.TypeDouble},
				ReturnType: ast.TypeDouble,
				CName:      "dex_math_ceil",
			},
			"round": {
				Params:     []ast.Type{ast.TypeDouble},
				ReturnType: ast.TypeDouble,
				CName:      "dex_math_round",
			},
			// Other
			"abs": {
				Params:     []ast.Type{ast.TypeDouble},
				ReturnType: ast.TypeDouble,
				CName:      "dex_math_abs",
			},
			"log": {
				Params:     []ast.Type{ast.TypeDouble},
				ReturnType: ast.TypeDouble,
				CName:      "dex_math_log",
			},
			"log2": {
				Params:     []ast.Type{ast.TypeDouble},
				ReturnType: ast.TypeDouble,
				CName:      "dex_math_log2",
			},
			"log10": {
				Params:     []ast.Type{ast.TypeDouble},
				ReturnType: ast.TypeDouble,
				CName:      "dex_math_log10",
			},
			// Two-arg
			"min": {
				Params:     []ast.Type{ast.TypeDouble, ast.TypeDouble},
				ReturnType: ast.TypeDouble,
				CName:      "dex_math_min",
			},
			"max": {
				Params:     []ast.Type{ast.TypeDouble, ast.TypeDouble},
				ReturnType: ast.TypeDouble,
				CName:      "dex_math_max",
			},
		},
		CFlags: []string{"-lm"},
		CRuntime:  mathRuntime,
	})
}
